File: scripts/increment-build.js
const fs = require('fs');
const path = require('path');

const projectRoot = path.resolve(__dirname, '..');
const buildJsonPath = path.join(projectRoot, 'build.json');
const winresJsonPath = path.join(projectRoot, 'winres', 'winres.json');

function formatDate(d) {
  const pad = (n) => String(n).padStart(2, '0');
  const year = d.getFullYear();
  const month = pad(d.getMonth() + 1);
  const day = pad(d.getDate());
  const hours = pad(d.getHours());
  const minutes = pad(d.getMinutes());
  const seconds = pad(d.getSeconds());
  return `${year}-${month}-${day} ${hours}:${minutes}:${seconds}`;
}

function updateWinres(buildNum, buildDate) {
  if (!fs.existsSync(winresJsonPath)) return;
  try {
    const wdata = JSON.parse(fs.readFileSync(winresJsonPath, 'utf8'));
    const verStr = `1.0.0.${buildNum}`;

    if (wdata.RT_MANIFEST && wdata.RT_MANIFEST['#1'] && wdata.RT_MANIFEST['#1']['0409']) {
      const manifest = wdata.RT_MANIFEST['#1']['0409'];
      if (manifest.identity) manifest.identity.version = verStr;
    }

    if (wdata.RT_VERSION && wdata.RT_VERSION['#1'] && wdata.RT_VERSION['#1']['0000']) {
      const verObj = wdata.RT_VERSION['#1']['0000'];
      if (verObj.fixed) {
        verObj.fixed.file_version = verStr;
        verObj.fixed.product_version = verStr;
      }
      if (verObj.info && verObj.info['0409']) {
        const info = verObj.info['0409'];
        info.FileVersion = verStr;
        info.ProductVersion = verStr;
        info.Comments = `Build #${buildNum} (${buildDate})`;
      }
    }

    fs.writeFileSync(winresJsonPath, JSON.stringify(wdata, null, 2) + '\n', 'utf8');
  } catch (e) {
    console.warn('[Aviso] Nao foi possivel atualizar winres.json:', e.message);
  }
}

function incrementBuild() {
  let data = { build: 0, date: '' };
  if (fs.existsSync(buildJsonPath)) {
    try {
      data = JSON.parse(fs.readFileSync(buildJsonPath, 'utf8'));
    } catch (e) {
      data = { build: 0, date: '' };
    }
  }

  data.build = (parseInt(data.build, 10) || 0) + 1;
  data.date = formatDate(new Date());

  fs.writeFileSync(buildJsonPath, JSON.stringify(data, null, 2) + '\n', 'utf8');
  updateWinres(data.build, data.date);

  console.log(`[Build] Versao incrementada para Build #${data.build} (${data.date})`);
  return data;
}

if (require.main === module) {
  incrementBuild();
}

module.exports = { incrementBuild };

File: scripts/increment_build.py
import datetime
import json
import os

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
PROJECT_ROOT = os.path.abspath(os.path.join(SCRIPT_DIR, ".."))
BUILD_JSON_PATH = os.path.join(PROJECT_ROOT, "build.json")
WINRES_JSON_PATH = os.path.join(PROJECT_ROOT, "winres", "winres.json")


def format_date(d):
    return d.strftime("%Y-%m-%d %H:%M:%S")


def update_winres(build_num, build_date):
    if not os.path.exists(WINRES_JSON_PATH):
        return
    try:
        with open(WINRES_JSON_PATH, "r", encoding="utf-8") as f:
            wdata = json.load(f)

        ver_str = f"1.0.0.{build_num}"

        # Update Manifest identity version
        if "RT_MANIFEST" in wdata and "#1" in wdata["RT_MANIFEST"]:
            manifest = wdata["RT_MANIFEST"]["#1"].get("0409", {})
            if "identity" in manifest:
                manifest["identity"]["version"] = ver_str

        # Update RT_VERSION fixed and info
        if "RT_VERSION" in wdata and "#1" in wdata["RT_VERSION"]:
            ver_obj = wdata["RT_VERSION"]["#1"].get("0000", {})
            if "fixed" in ver_obj:
                ver_obj["fixed"]["file_version"] = ver_str
                ver_obj["fixed"]["product_version"] = ver_str
            if "info" in ver_obj and "0409" in ver_obj["info"]:
                info = ver_obj["info"]["0409"]
                info["FileVersion"] = ver_str
                info["ProductVersion"] = ver_str
                info["Comments"] = f"Build #{build_num} ({build_date})"

        with open(WINRES_JSON_PATH, "w", encoding="utf-8") as f:
            json.dump(wdata, f, indent=2)
            f.write("\n")
    except Exception as e:
        print(f"[Aviso] Nao foi possivel atualizar winres.json: {e}")


def increment_build():
    data = {"build": 0, "date": ""}
    if os.path.exists(BUILD_JSON_PATH):
        try:
            with open(BUILD_JSON_PATH, "r", encoding="utf-8") as f:
                data = json.load(f)
        except Exception:
            data = {"build": 0, "date": ""}

    curr_build = data.get("build", 0)
    try:
        curr_build = int(curr_build)
    except (ValueError, TypeError):
        curr_build = 0

    data["build"] = curr_build + 1
    data["date"] = format_date(datetime.datetime.now())

    with open(BUILD_JSON_PATH, "w", encoding="utf-8") as f:
        json.dump(data, f, indent=2)
        f.write("\n")

    update_winres(data["build"], data["date"])

    print(f"[Build] Versao incrementada para Build #{data['build']} ({data['date']})")
    return data


if __name__ == "__main__":
    increment_build()
